### ISSUE TITLE
[feature request] Autocast module wrappers

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#70166]
### 🚀 The feature, motivation and pitch

From what I understand from https://pytorch.org/docs/stable/amp.html, decorators are currently only supported on forward method. For composability, it's sometimes useful to have an autocast decorator called on module object, similar to https://github.com/pytorch/pytorch/pull/70164/files (instead of using autocast with blocks). When this happens, original module code doesn't need to be modified and it's useful while researching feasibility of autocast of various model components without having to spill around with autocast blocks.

Some more context: https://discuss.pytorch.org/t/can-autocast-context-manager-be-used-around-all-of-training-loop/136771/7?u=vadimkantorov

**Comment by vadimkantorov** [comment]
Might be more clear what's going on with autocast x export interaction if we could enable / control autocast with a simple module object wrappers  (context: https://github.com/pytorch/pytorch/issues/72494)

**Comment by vadimkantorov** [comment]
Somewhat related: https://github.com/pytorch/pytorch/issues/75478 - in the sense of forcing full/mixed precision for certain modules. Similar decorators exist in mmdet/mmcv: https://mmcv.readthedocs.io/en/latest/api.html#mmcv.runner.force_fp32

**Comment by vadimkantorov** [comment]
Somewhat related: https://github.com/pytorch/pytorch/issues/76578

**Comment by vadimkantorov** [comment]
Maybe related: https://github.com/pytorch/pytorch/issues/79250

**Comment by rohan-varma** [comment]
This is actually a pretty cool idea - do you mean instead of having

```
with autocast():
    model(inputs)
```

we do something like
```
module_with_autocast = wrap_with_autocast(module)
```?

**Comment by vadimkantorov** [comment]
Yes, either: `module = wrap_module_instance_with_autocast(module)` or `MyModuleAutocasted = wrap_module_class_with_autocast(MyModule)`. Possibly also a `my_function = wrap_function_with_autocast(my_function, "f16", "f32")` could be considered.

All of these make it much easier to reason/debug/hack/improve about how actually autocast works, what gets casted at what time, how to exempt some parts from autocast and how all this interacts with other global cross-cutting aspects such as training loop or export/scripting

At least in many cases, there is value in this (although probably it can make graph-matching+replacement harder)

**Comment by vadimkantorov** [comment]
Probably, tracing would be simpler if the autocast could explicitly represented as modules: https://github.com/pytorch/pytorch/issues/84092

**Comment by rohan-varma** [comment]
Autocast wrapper sounds good to me. Wonder who's the owner of autocast to provide feedback and take this idea forward, any ideas @ezyang @ngimel?

**Comment by ezyang** [comment]
@zou3519 PyOperators would give an api for doing this
